docs/ldml/tr35-keyboards.md

```diff
@@ -308,6 +308,336 @@ For extensibility, the `<special>` element will be allowed at nearly every level
+
+For example, the sequence `e\m{marker}\u{300}` would be output in NFC as `è`. The marker is removed and has no effect on output.
+
+### Normalization-safe Segments
+
+For purposes of the above discussion, "normalization-safe segments" are defined as a string of codepoints which are
+
+1. already in [NFD](https://www.unicode.org/reports/tr15/#Norm_Forms), and
+2. begin with a character with [Canonical Combining Class](https://www.unicode.org/reports/tr44/#Canonical_Combining_Class_Values) of `0`.
+
+See [UAX #15 Section 9.1: Stable Code Points](https://www.unicode.org/reports/tr15/#Stable_Code_Points) for related discussion.
+Text under consideration can be segmented by locating such characters.
+
+### Pseudocode for Marker-safe Normalization
+
+The following is an informative section with some example pseudocode for marker-safe normalization into NFD.
+
+1. Markers are not text. This pseudocode assumes an implementation which is able to distinguish markers from plain text.
+
+2. Some optimization notes are given, but the emphasis here is on a clear and correct implementation.
+
+3. A distinction would need to be made between (A) markers in text, such as a keystroke emitting `\m{someMarker}` or a marker in input context, and (B) markers in transform match strings, such as `<transform from="\m{marker}">` or `<transform from="\m{.}">`.  An implementation could pass a parameter through the following code path distinguishing A and B, this distinction is not shown, for clarity.
+
+Example use:
+
+- `nfdMarkers('e\u{0300}' + Marker('marker1') + '\u{0320}a\u{0300}' + Marker('marker2') + '\u{0320}')`
+- Returns: `'e' + Marker('marker1') + '\u{0320}\u{0300}a' + Marker('marker2') + '\u{0320}\u{0300}'`
+
+```js
+/**
+ * Normalize a string-with-markers into NFD.
+ * @param input string with markers as out-of-band
+ * @returns normalized output as a string with markers
+ */
+export function nfdMarkers(input /* string with markers */) {
+    // to accumulate output text
+    let output = "";
+    // remove all markers
+    let noMarkers = removeMarkers(input);
+    // NFD the no-marker text.
+    let noMarkersNfd = normalizeNFD(noMarkers);
+
+    // Two quick checks.
+    if (noMarkers === input) {
+        // Heuristic: if there were no markers we can safely return noMarkersNFD
+        return noMarkersNFD;
+    } else if (noMarkersNFD === noMarkers) {
+        // Heuristic: if normalization made no change, return the input string: already NFD
+        return input;
+    }
+
+    let segmentStart = 0;
+    let segmentEnd = 0;
+
+    // look for normalization-safe segments
+    // 'i' = iterate by codepoints
+    do {
+        const haveMarker = markerAt(i); // true if there is a marker here
+        if (i === input.end) {
+            segmentEnd = i; // end of input - end of segment
+        } else if(hasNFDBoundaryBefore(i) && !have_marker) {
+            segmentEnd = i; // end of segment
+            i++; // move past the NFD char
```

Comment:
Using js conventions (and \`\`\`js)can make people think this is mostly js. But js (and java) will fail with i++ (here and elsewhere).